internal/snapshots/snapshotter.go

```diff
@@ -0,0 +1,444 @@
+	// create snapshot directory
+	snapshotDir := SnapshotHeightDir(s.snapshotDir, height)
+	chunkDir := SnapshotChunkDir(s.snapshotDir, height, DefaultSnapshotFormat)
+	err := os.MkdirAll(chunkDir, 0755)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create snapshot chunk dir: %w", err)
+	}
+
+	// Stage1: Dump the database at the given height and snapshot ID
+	err = s.dbSnapshot(ctx, height, DefaultSnapshotFormat, snapshotID)
+	if err != nil {
+		deleteSnapshotDir(snapshotDir)
+		return nil, fmt.Errorf("failed to create snapshot at height %d: %w", height, err)
+	}
+
+	// Stage2: Sanitize the dump
+	hash, err := s.sanitizeDump(height, DefaultSnapshotFormat)
+	if err != nil {
+		deleteSnapshotDir(snapshotDir)
+		return nil, fmt.Errorf("failed to sanitize snapshot at height %d: %w", height, err)
+	}
+
+	// Stage3: Compress the dump
+	err = s.compressDump(height, DefaultSnapshotFormat)
+	if err != nil {
+		deleteSnapshotDir(snapshotDir)
+		return nil, fmt.Errorf("failed to compress snapshot at height %d: %w", height, err)
+	}
+
+	// Stage4: Split the dump into chunks
+	snapshot, err := s.splitDumpIntoChunks(height, DefaultSnapshotFormat, hash)
+	if err != nil {
+		deleteSnapshotDir(snapshotDir)
+		return nil, fmt.Errorf("failed to split snapshot into chunks at height %d: %w", height, err)
+	}
+
+	return snapshot, nil
+}
+
+// dbSnapshot is the STAGE1 of the snapshot creation process
+// It uses pg_dump to dump the database state at the given height and snapshotID
+// The pg dump is stored as "/stage1output.sql" in the snapshot directory
+// This is a temporary file and will be removed after the snapshot is created
+func (s *Snapshotter) dbSnapshot(ctx context.Context, height uint64, format uint32, snapshotID string) error {
+	snapshotDir := SnapshotFormatDir(s.snapshotDir, height, format)
+	dumpFile := filepath.Join(snapshotDir, stage1output)
+
+	pgDumpCmd := exec.CommandContext(ctx,
+		"pg_dump",
+		// File format options
+		"--dbname", "kwild",
+		"--file", dumpFile,
+		"--format", "plain",
+		// List of schemas to include in the dump
+		"--schema", "kwild_voting",
+		"--schema", "kwild_chain",
+		"--schema", "kwild_accts",
+		"--schema", "kwild_internal",
+		"--schema", "ds_*",
+		// Tables to exclude from the dump
+		"-T", "kwild_internal.sentry",
+		// other sql dump specific options
+		"--no-unlogged-table-data",
+		"--no-comments",
+		"--create",
+		"--no-publications",
+		"--no-unlogged-table-data",
+		"--no-tablespaces",
+		"--no-table-access-method",
+		"--no-security-labels",
+		"--no-subscriptions",
+		"--large-objects",
+		// Snapshot ID ensures a consistent snapshot taken at the given block boundary across all nodes
+		"--snapshot", snapshotID,
+		// Connection options
+		"-U", s.dbConfig.DBUser,
+		"-h", s.dbConfig.DBHost,
+		"-p", s.dbConfig.DBPort,
+	)
+
+	s.log.Debug("Executing pg_dump", zap.String("cmd", pgDumpCmd.String()))
+
+	output, err := pgDumpCmd.CombinedOutput()
```

Comment:
I'm a bit concerned about running out of memory here. We are essentially dumping the entire DB to memory, right?